### ISSUE TITLE
SQLSTATE[HY000] [2002] Connection refused

### ISSUE THREAD

**Issue by omag0** [laradock/laradock#1323]
- System info ( Linux): 
 

Hello guys. Have a problem, and need help.

I clone laradock in /var/www/docker/laradock
than i make docker-compose up -d apache2 mysql

and in /var/www/docker i run laravel new project.

my laradock .env
MYSQL_VERSION=8.0
MYSQL_DATABASE=AriesDB
MYSQL_USER=default
MYSQL_PASSWORD=secret
MYSQL_PORT=3307
MYSQL_ROOT_PASSWORD=root
MYSQL_ENTRYPOINT_INITDB=./mysql/docker-entrypoint-initdb.d

my laravel .env

DB_CONNECTION=mysql
DB_HOST=127.0.0.1
DB_PORT=3307
DB_DATABASE=AriesDB
DB_USERNAME=root
DB_PASSWORD=root

when i do "php artisan migrate" and "php artisan db:seed" everything works ok
buy when i run my browser and put my login and password, show this messenger SQLSTATE[HY000] [2002] Connection refused (SQL: select * from `users` where `email` = rodrigo@com.br limit 1).

if i do docker-compose exec workspace bash, and try "php artisan migrate" show this messenger: SQLSTATE[HY000] [2002] Connection refused  

I tryied change DB_HOST=127.0.0.1 to DB_HOST=mysql but nothing changed.
Can somebody help me?

**Comment by lcdss** [comment]
Have you tried to change the MYSQL_HOST to `mysql` or `mariadb`, depending on which database container you are using? Is the port 3307 correct as well?

**Comment by omag0** [comment]
Yep, i tried. And is the port 3307.

Without using laravel, and using just using pdo, the connection works fine.

But with laravel / eloquent, nothing works.....

Im lost....

________________________________
De: Lucas C. S. Silva <notifications@github.com>
Enviado: quarta-feira, 27 de dezembro de 2017 22:50:47
Para: laradock/laradock
Cc: Rodrigo Batista; Author
Assunto: Re: [laradock/laradock] SQLSTATE[HY000] [2002] Connection refused (#1323)


Have you tried to change the MYSQL_HOST to mysql or mariadb, depending on which database container you are using? Is the port 3307 correct as well?

**Comment by harrysbaraini** [comment]
Use DB_HOST=mysql instead of 127.0.0.1. When running outside of workspace container, the MySQL container is accessible on 127.0.0.1, but when you're in the workspace container, 127.001 is relative to the container.

If you want to run php artisan commands outside of the workspace container, you can set DB_HOST=<<your docker ip>>, but I don't recommend so, because your environment may be different of the containers.

**Comment by ed-fruty** [comment]
@omag0 when you want to use cli commands, like `artisan` you must enter into the `workspace` container, like

```
cd /var/www/docker/laradock
docker-compose exec --user=laradock workspace bash

# here you can use cli
php artisan migrate
```

**Comment by inthedeepend** [comment]
@harrysbaraini answer is correct.

In simpler terms : 
```
DB_CONNECTION=mysql
DB_HOST=mysql
DB_PORT=3306
DB_DATABASE=db
DB_USERNAME=username
DB_PASSWORD=password
```

**Comment by rafa-acioly** [comment]
I'm facing this problem as well, but i've already config. my `.env` file:

```
DB_CONNECTION=mysql
DB_HOST=mysql
DB_PORT=3306
DB_DATABASE=mydb
DB_USERNAME=root
DB_PASSWORD=root
```

> PDOException::("SQLSTATE[HY000] [1130] Host '172.20.0.3' is not allowed to connect to this MySQL server")
      /var/www/ftimesheet/vendor/laravel/framework/src/Illuminate/Database/Connectors/Connector.php : 68

>PDO::__construct("mysql:host=mysql;port=3306;dbname=mydb", "root", "root", [])
      /var/www/ftimesheet/vendor/laravel/framework/src/Illuminate/Database/Connectors/Connector.php : 68

**Comment by danaia** [comment]
I am getting the exact same error as @rafa-acioly with the same .env file. This has been running fine for a few months. Made no changes to the Laradock files Getting this error from a clean install and running from a Laradock fork which has not failed me until now.

**Comment by tpraxl** [comment]
Just had the same problem. Turned out that this was due to a misconfiguration.

I think I'll post  a complete example, might help others to understand the issue.

Given you have the following section in your laradock .env

```
MYSQL_VERSION=latest
MYSQL_DATABASE=default
MYSQL_USER=default
MYSQL_PASSWORD=secret
MYSQL_PORT=3307
MYSQL_ROOT_PASSWORD=root
MYSQL_ENTRYPOINT_INITDB=./mysql/docker-entrypoint-initdb.d
```
Then you need the following laravel .env

```
DB_CONNECTION=mysql
DB_HOST=mysql
DB_PORT=3306
DB_DATABASE=default
DB_USERNAME=default
DB_PASSWORD=secret
```
Then you can `docker-compose exec --user=laradock workspace bash` and 

``` 
php artisan migrate:install
php artisan migrate
```

Wondering why `DB_PORT=3306` is configured in laravel's .env?

Actually, the port you specified in laradock's .env is the port mapped to the host. It should not matter for container communication. 

## Guessing what to do if you wanted to change the port for containers

I guess, if you wanted to change the mysql port for container communication, you would need to change mysql/my.cnf 

```
[mysql]
port=3309
```

And you might need to change the mysql/Dockerfile to `EXPOSE 3309` then. You might also need to adapt the docker-compose.yml to this:

```
ports:
        - "${MYSQL_PORT}:3309"
```

**Comment by NameIvan** [comment]
Use container name of mysql in DB_HOST. Like this DB_HOST=laravellocal_mysql_1

**Comment by Lewi77** [comment]
If u have changed the `MYSQL_PORT` in  laradock `.env` , u may changed the `DB_PORT` in laravel `.env` too. 
So, just keep the `DB_PORT=3306` in laravel `.env`. I have solved it in this way.
![image](https://user-images.githubusercontent.com/16359506/61107371-55ac4d00-a4b2-11e9-8328-b03a8ab3caf6.png)

**Comment by phitran35** [comment]
> Use `DB_HOST=mysql` instead of 127.0.0.1. When running outside of workspace container, the MySQL container is accessible on 127.0.0.1, but when you're in the workspace container, 127.001 is relative to the container itself, so container will try to connect to its own localhost.
> 
> as @inthedeepend said below, your .env should look like this:
> 
> ```
> DB_CONNECTION=mysql
> DB_HOST=mysql
> DB_PORT=3306
> DB_DATABASE=db
> DB_USERNAME=username
> DB_PASSWORD=password
> ```
> 
> If you want to run php artisan commands outside of the workspace container, you can set `DB_HOST=your docker ip`, but I don't recommend so, because your environment may be different of the containers, e.g., have some PHP library missing.

I have to say thanks to. you. Love it

**Comment by Dazeh** [comment]
Anyone still facing this issue - If you ran `php artisan config:cache` on your local machine (in my case windows), you instead need to run the artisan command from within your workspace container as this is creating a config file which has your windows paths instead of the containers
